### ISSUE TITLE
[web]: Flicker issue when using HtmlElementView

### ISSUE THREAD

**Issue by rohansohonee1** [flutter/flutter#150979]
### Steps to reproduce

1. Run the sample code for the web in release mode.
2. Observe the flicker/blank page in output.
3. Run the sample code for the web in debug mode.
4. Observe the web page is loaded fine & there is no flicker/blank page issue.
5. Run the sample code for the web in debug mode once again with a small change.
5a. Add the `debugShowCheckedModeBanner: false` to the MaterialApp.
5b. Observe the flicker/blank page in output.

### Expected results

The web page should be loaded fine without any flicker/blank page issues.

### Actual results

The web page is getting flicker/blank issue in release mode & when setting `debugShowCheckedModeBanner: false` in MaterialApp.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'dart:js_interop';
import 'dart:ui_web' as ui_web;

import 'package:flutter/material.dart';
import 'package:web/web.dart' as web;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final element = web.document.createElement("iFrame") as web.HTMLIFrameElement;

  final String viewTypeId = "12345";

  @override
  void initState() {
    super.initState();

    // add the message event listener
    web.window.addEventListener('message', messageEvent.toJS, true.toJS);

    // set up iFrame element
    element
      ..style.border = '0'
      ..allowFullscreen = false
      ..id = "iFrameElement-$viewTypeId"
      ..height = '100%'
      ..width = '100%';

    String url = "https://flutter.dev";
    debugPrint(url);
    element.src = url;

    element.onLoad.listen((event) {
      debugPrint("iFrame loaded");
    });

    ui_web.platformViewRegistry
        .registerViewFactory('iFrame-$viewTypeId', (int viewId) => element);
  }

  @override
  void dispose() {
    // remove the message event listener
    web.window.removeEventListener('message', messageEvent.toJS, true.toJS);
    element.removeAttribute("src");
    super.dispose();
  }

  /// callback for "message" event listener
  void messageEvent(web.Event e) {
    if (!mounted) return;

    var data = (e as web.MessageEvent).data?.dartify() as Map? ?? {};
    debugPrint("Message Event data >>>> $data");
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: [
          AppBar(title: const Text("Flicker issue")),
          Expanded(
            child: AbsorbPointer(
              child: RepaintBoundary(
                child: HtmlElementView(
                  key: widget.key,
                  viewType: 'iFrame-$viewTypeId',
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

### Release mode output:
<img width="1257" alt="Screen Shot 2024-06-28 at 4 38 30 PM" src="https://github.com/flutter/flutter/assets/120708465/50757060-93e9-482f-9b22-f18844063aff">

### Debug mode output:
<img width="1257" alt="Screen Shot 2024-06-28 at 4 39 28 PM" src="https://github.com/flutter/flutter/assets/120708465/05bd149d-8993-4994-8c52-e1704a7d51c2">

</details>

### Logs

<details open><summary>Logs</summary>

```console
Launching lib/main.dart on Chrome in debug mode...
Waiting for connection from debug service on Chrome...             21.5s
This app is linked to the debug service: ws://127.0.0.1:63517/e0mnCKGbWg0=/ws
Debug service listening on ws://127.0.0.1:63517/e0mnCKGbWg0=/ws

🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".

A Dart VM Service on Chrome is available at: http://127.0.0.1:63517/e0mnCKGbWg0=
https://flutter.dev
Height of Platform View type: [iFrame-12345] may not be set. Defaulting to `height: 100%`.
Set `style.height` to any appropriate value to stop this message.
Width of Platform View type: [iFrame-12345] may not be set. Defaulting to `width: 100%`.
Set `style.width` to any appropriate value to stop this message.
The Flutter DevTools debugger and profiler on Chrome is available at: http://127.0.0.1:9101?uri=http://127.0.0.1:63517/e0mnCKGbWg0=
iFrame loaded
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.19.6, on macOS 12.7.5 21H1222 darwin-x64, locale en-US)
    • Flutter version 3.19.6 on channel stable at /Users/appleapple/fvm/versions/3.19.6
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 54e66469a9 (2 months ago), 2024-04-17 13:08:03 -0700
    • Engine revision c4cd48e186
    • Dart version 3.3.4
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/appleapple/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] VS Code (version 1.90.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.7.5 21H1222 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 126.0.6478.127

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by rohansohonee1** [comment]
We cannot upgrade our project yet to Flutter version 3.22.2 (latest stable).

**Comment by rohansohonee1** [comment]
I have tested this sample code with the Flutter version 3.22.2 (latest stable).
The issue seems to be resolved.

Can you please share the PR in which this issue was addressed?
I have searched in these links but was not able to find the exact fix:
- https://docs.flutter.dev/release/release-notes/release-notes-3.22.0
- https://github.com/flutter/flutter/blob/master/docs/releases/Hotfixes-to-the-Stable-Channel.md

**Comment by rohansohonee1** [comment]
I have tested this sample code with the Flutter version 3.22.2 (latest stable).
The issue seems to be resolved.

Can you please share the PR in which this issue was addressed?
I have searched in these links but was not able to find the exact fix:
- https://docs.flutter.dev/release/release-notes/release-notes-3.22.0
- https://github.com/flutter/flutter/blob/master/docs/releases/Hotfixes-to-the-Stable-Channel.md

**Comment by darshankawar** [comment]
@rohansohonee1 
There have been few fixes made to HtlmlElementView like below that you can check and see:

https://github.com/flutter/flutter/issues/143922
https://github.com/flutter/flutter/issues/149983#issuecomment-2173970887

**Comment by rohansohonee1** [comment]
Can you share a workaround for this issue when using Flutter version 3.19.6? Because we cannot upgrade our project yet to 3.22.2

**Comment by darshankawar** [comment]
I am not sure about the workaround. Looking at the code, you seem to be already passing 100% height and width.
I'll keep this issue open for team's input for workaround, although in latest version (3.22.2), it doesn't flicker as reported.

/cc @ditman for thoughts.

**Comment by ditman** [comment]
@darshankawar I haven't been able to reproduce this issue in 3.19.6, using @rohansohonee1's repro code:

https://github.com/flutter/flutter/assets/1255594/9901c9f2-9746-44ce-beb3-fbc864910fd5

The issues with flickering of platform views may be related to this family of fixes:

* https://github.com/flutter/engine/pull/47317

Unfortunately, without identifying the issue it's hard to offer a workaround :S

**Comment by ditman** [comment]
In order to move forward we need more detailed information from @rohansohonee1 (Version info on the Browser, iOS, Flutter... errors in the JS console of the browser); a reliable way to reproduce the issue.

Assigning this a low priority, because the issue is [already fixed in `stable`](https://github.com/flutter/flutter/issues/150979#issuecomment-2196678928).

**Comment by rohansohonee1** [comment]
@ditman First close the chrome inspect window. Then run the sample code. In my testing I also found that it worked when I opened the chrome inspect window. I will share a video of the issue shortly.

I have already shared detailed information in Logs & Flutter doctor output section.

**Comment by ditman** [comment]
> I have already shared detailed information in Logs & Flutter doctor output section.

@rohansohonee1 you forgot your versions of MacOS (14.2?) and Chrome (126?), but I found that in some of the videos, so it's all right.

----

Unfortunately, with Chrome 126.0.6478.127, I can't reproduce neither on gLinux:

https://github.com/flutter/flutter/assets/1255594/7e3817d1-84ba-4624-ac9d-8587bcd9232c

nor **Mac OS (Sonoma 14.5)**:

https://github.com/flutter/flutter/assets/1255594/f83693b1-3879-4d04-a01b-52f343475f70

with the source code you shared above.

I've noticed that your version of Mac OS is also a little bit old (12.4?), so this might be an Chrome + old(ish) Mac OS issue. If this were the case, I wouldn't be able to downgrade my work computer to verify (infosec policy).

> Can you share a workaround for this issue when using Flutter version 3.19.6? Because we cannot upgrade our project yet to 3.22.2

@rohansohonee1 as a workaround (and if this were related to the layer optimizations in the engine), I'd suggest you try to add something similar to the `debugShowCheckedModeBanner` to your app. I don't think it needs to be huge, just barely visible (1x1px) to maybe fix the issue, at least until you manage to upgrade to the next stable. If it's not related to the layer optimization, I don't know.

---

@darshankawar, @rohansohonee1, there isn't much else I can do here, so I'm going to close the issue. If anybody else comes up with a reproduction case, I'd love to take a look, but this is about a month away to becoming "2 stable versions ago", and I need to focus in fixing things that are still broken.
